bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java

```diff
@@ -228,28 +232,67 @@ boolean doRefresh() {
     if (!requiresRefresh) {
       return false;
     }
-    HeaderCacheElement cacheElement = null;
+    HeaderCacheElement cacheElement = refreshCredentialsWithRetry(null);
+    synchronized (isRefreshing) {
+      headerCache.set(cacheElement);
+      isRefreshing.set(false);
+      isRefreshing.notifyAll();
+    }
+    return true;
+  }
+
+
+  protected HeaderCacheElement refreshCredentialsWithRetry(BackOff backoff) {
     try {
       AccessToken newToken = credentials.refreshAccessToken();
       if (newToken == null) {
-        // The current implementations of refreshAccessToken() throw an IOException or return
-        // a valid token. This handling is future proofing.
-        cacheElement = new HeaderCacheElement(
+        // The current implementations of refreshAccessToken() throws an IOException or return
+        // a valid token. This handling is future proofing just in case credentials returns null for
+        // some reason.  This case was caught by a poorly coded unit test.
+        return new HeaderCacheElement(
             new IOException("Could not load the token for credentials: " + credentials));
       } else {
-        cacheElement = new HeaderCacheElement(newToken);
+        // Success!
+        return new HeaderCacheElement(newToken);
       }
-    } catch (IOException e) {
-      cacheElement = new HeaderCacheElement(e);
+    } catch (IOException exception) {
+      // An IOException occurred.  Retry with backoff.
+      if (backoff == null) {
+        // initialize backoff.
+        backoff = retryOptions.createBackoff();
+      }
+
+      // Given the backoff, either sleep for a short duration, or terminate if the backoff has
+      // reached its configured timeout limit.
+      boolean isTimedOut = sleepOrTerminate(backoff);
+      return isTimedOut ? new HeaderCacheElement(exception) : refreshCredentialsWithRetry(backoff);
     } catch (Exception e) {
-      cacheElement = new HeaderCacheElement(new IOException("Could not read headers", e));
-    } finally {
-      synchronized (isRefreshing) {
-        headerCache.set(cacheElement);
-        isRefreshing.set(false);
-        isRefreshing.notifyAll();
+      return new HeaderCacheElement(new IOException("Could not read headers", e));
+    }
+
+  }
+
+  /**
+   * Sleep and/or determine if the backoff has timed out.
+   * @param backoff
+   * @return boolean indicating whether or not to terminate retries. True = terminate, False =
+   *         continue.
+   */
+  protected boolean sleepOrTerminate(BackOff backoff) {
+    long nextBackOffMillis = 0;
+    try {
+      nextBackOffMillis = backoff.nextBackOffMillis();
```

Comment:
Maybe we can clean this up a bit:
long nextBackOffMillis = backoff.nextBackOffMillis();
if (nextBackOffMillis == BackOff.STOP) {
  return true;
}
try { ... }